sparse/_sparse_array.py

```diff
@@ -259,3 +264,436 @@ def __array_function__(self, func, types, args, kwargs):
+        keepdims : bool, optional
+            Whether or not to keep the dimensions of the original array.
+        kwargs : dict
+            Any extra arguments to pass to the reduction operation.
+
+        Returns
+        -------
+        GCXS
+            The result of the reduction operation.
+
+        Raises
+        ------
+        ValueError
+            If reducing an all-zero axis would produce a nonzero result.
+
+        See Also
+        --------
+        numpy.ufunc.reduce : A similar Numpy method.
+        COO.reduce : Equivalent operation on COO arrays.
+
+        Examples
+        --------
+        You can use the :obj:`COO.reduce` method to apply a reduction operation to
+        any Numpy :code:`ufunc`.
+
+        >>> from sparse import COO
+        >>> x = np.ones((5, 5), dtype=np.int)
+        >>> s = COO.from_numpy(x)
+        >>> s2 = s.reduce(np.add, axis=1)
+        >>> s2.todense()  # doctest: +NORMALIZE_WHITESPACE
+        array([5, 5, 5, 5, 5])
+
+        You can also use the :code:`keepdims` argument to keep the dimensions after the
+        reduction.
+
+        >>> s3 = s.reduce(np.add, axis=1, keepdims=True)
+        >>> s3.shape
+        (5, 1)
+
+        You can also pass in any keyword argument that :obj:`numpy.ufunc.reduce` supports.
+        For example, :code:`dtype`. Note that :code:`out` isn't supported.
+
+        >>> s4 = s.reduce(np.add, axis=1, dtype=np.float16)
+        >>> s4.dtype
+        dtype('float16')
+
+        By default, this reduces the array by only the first axis.
+
+        >>> s.reduce(np.add)
+        <COO: shape=(5,), dtype=int64, nnz=5, fill_value=0>
+        """
+        axis = normalize_axis(axis, self.ndim)
+        zero_reduce_result = method.reduce([self.fill_value, self.fill_value], **kwargs)
+        reduce_super_ufunc = None
+
+        if not equivalent(zero_reduce_result, self.fill_value):
+            reduce_super_ufunc = _reduce_super_ufunc.get(method, None)
+
+            if reduce_super_ufunc is None:
+                raise ValueError(
+                    "Performing this reduction operation would produce "
+                    "a dense result: %s" % str(method)
+                )
+
+        if axis is None:
+            if hasattr(self, "indptr"):  # gcxs
+                if reduce_super_ufunc is None:
+                    temp = method(self.data, self.fill_value)
+                    val = method.reduce(temp, **kwargs)
+                else:
+                    temp = method(
+                        self.data,
+                        reduce_super_ufunc(self.fill_value, self.size - self.nnz),
+                    )
+                    val = method.reduce(temp, **kwargs)
+                if keepdims:
+                    out = np.array(val)
+                    return out.reshape(np.ones(self.ndim, dtype=np.intp))
+                return val
+            else:  # coo
+                axis = tuple(range(self.ndim))
```

Comment:
Can we move this to a method that does `raise NotImplementedError` on `SparseArray` and implement it on `COO` and `GCXS`